File: 0317-shortest-distance-from-all-buildings/0317-shortest-distance-from-all-buildings.py
class Solution:
    def shortestDistance(self, g: List[List[int]]) -> int:
        n, m = len(g), len(g[0])
        distance = [[0] * m for _ in range(n)]
        reach = [[0] * m for _ in range(n)]
        buildings = 0 
        for r in range(n):
            for c in range(m):
                if g[r][c] == 1:
                    vis = set()

                    buildings += 1
                    q = deque([(r, c, 0)])
                    vis.add((r, c))
                    while q:
                        i, j, d = q.popleft()
                        for dx, dy in [(0,1), (1,0), (-1, 0), (0,-1)]:
                            x, y = i + dx, j + dy
                            if x >= 0 and y >= 0 and x < n and y < m and g[x][y] == 0 and (x,y) not in vis:
                                vis.add((x, y))
                                distance[x][y] += 1 + d
                                q.append((x, y, 1 + d))
                                reach[x][y] += 1

                    vis.clear()

        shortes = inf
        for i in range(n):
            for j in range(m):
                if reach[i][j] == buildings:
                    shortes = min(shortes, distance[i][j])

        return shortes if shortes != inf else -1

File: 0317-shortest-distance-from-all-buildings/0317-shortest-distance-from-all-buildings.java
import java.util.*;

class Solution {
    public int shortestDistance(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        
        // Accumulates total distance from all buildings for each cell
        int[][] distance = new int[n][m];
        // Tracks how many buildings can reach each cell
        int[][] reach = new int[n][m];
        int totalBuildings = 0;
        
        int[][] directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < m; c++) {
                if (grid[r][c] == 1) {
                    totalBuildings++;
                    bfs(grid, r, c, distance, reach, directions);
                }
            }
        }
        
        int shortest = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                // Check if the cell is empty land and can reach all buildings
                if (grid[i][j] == 0 && reach[i][j] == totalBuildings) {
                    shortest = Math.min(shortest, distance[i][j]);
                }
            }
        }
        
        return shortest == Integer.MAX_VALUE ? -1 : shortest;
    }
    
    private void bfs(int[][] grid, int r, int c, int[][] distance, int[][] reach, int[][] directions) {
        int n = grid.length;
        int m = grid[0].length;
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];
        
        q.add(new int[]{r, c, 0});
        visited[r][c] = true;
        
        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int currR = curr[0];
            int currC = curr[1];
            int d = curr[2];
            
            for (int[] dir : directions) {
                int x = currR + dir[0];
                int y = currC + dir[1];
                
                // Only traverse through empty land (0)
                if (x >= 0 && x < n && y >= 0 && y < m && grid[x][y] == 0 && !visited[x][y]) {
                    visited[x][y] = true;
                    distance[x][y] += d + 1;
                    reach[x][y]++;
                    q.add(new int[]{x, y, d + 1});
                }
            }
        }
    }
}